### ISSUE TITLE
DbgEng crashes on UAF in worker thread

### ISSUE THREAD

**Issue by CouleeApps** [Vector35/debugger#545]
Tried starting the debugger and closing the tab. Seems like an inconsistent race condition between the worker thread trying to run the debugger and the ui freeing the adapter. Likely need to keep a ref to the adapter in the thread.

Repro steps:
1. Open file
2. Start debugger
3. Close file
4. ???
5. Crash (but only sometimes)

```
0:023> k
 # Child-SP          RetAddr               Call Site
00 000000a0`578ff740 00007ff8`94a0f7f3     debuggercore!BinaryNinjaDebugger::DbgEngAdapter::EngineLoop+0x48f [c:\jenkins\workspace\binaryninja\os\win64\public\debugger\core\adapters\dbgengadapter.cpp @ 590] 
01 (Inline Function) --------`--------     debuggercore!BinaryNinjaDebugger::DbgEngAdapter::ExecuteWithArgs::__l2::<lambda_19080241d285197b7d2857382d7da79e>::operator()+0x4a [c:\jenkins\workspace\binaryninja\os\win64\public\debugger\core\adapters\dbgengadapter.cpp @ 405] 
02 (Inline Function) --------`--------     debuggercore!std::invoke+0x4a [C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.34.31933\include\type_traits @ 1562] 
03 000000a0`578ffa40 00007ff8`bd1b0e72     debuggercore!std::thread::_Invoke<std::tuple<<lambda_19080241d285197b7d2857382d7da79e> >,0>+0x53 [C:\Program Files\Microsoft Visual Studio\2022\Professional\VC\Tools\MSVC\14.34.31933\include\thread @ 57] 
04 000000a0`578ffa80 00007ff8`befa7bd4     ucrtbase!thread_start<unsigned int (__cdecl*)(void *),1>+0x42
05 000000a0`578ffab0 00007ff8`bf24ced1     KERNEL32!BaseThreadInitThunk+0x14
06 000000a0`578ffae0 00000000`00000000     ntdll!RtlUserThreadStart+0x21
```

**Comment by CouleeApps** [comment]
So the steps that i took to get this to repro:

1. Open exe in binja
2. Start debugging
3. New tab
4. Tear tab with analysis into a new window
5. Close tab with analysis
6. Close window which had that tab
7. Windbg pops up

**Comment by CouleeApps** [comment]
Repro'd two more times in the same place, so I think these steps are decently consistent

**Comment by xusheng6** [comment]
another repro:

1. open a file
2. start debugging
3. split the to a new window
4. close the new window
5. when asked about the running target, click "yes" to stop it
6. in the original windows, try to launch the target
7. crash

**Comment by CouleeApps** [comment]
Still reproduces on 4.0.4896
<img width="1098" alt="image" src="https://github.com/Vector35/debugger/assets/711973/21ef204d-0074-4e2f-b021-53f123aba7be">

**Comment by xusheng6** [comment]
I managed to get a local repro. The only difference is that, for the newly split view, I must directly close the BN window rather than the tab.

The dynamics of the crash is:

1. The code in UINotification::OnBeforeCloseFile detects this is the last tab of the file, and the user instructs it to terminate the target. So it calls QuitAndWait() on the controller: https://github.com/Vector35/debugger/blob/ee1d1c51dc99068645559184fb7fbde516c5f174/ui/uinotification.cpp#L126
2. The controller properly terminates the target, and get the `DEBUG_STATUS_NO_DEBUGGEE` status, which it proceeds with a `PostDebuggerEvent`. https://github.com/Vector35/debugger/blob/318e61219ced0815351eea16c6388d3e1372e6f7/core/adapters/dbgengadapter.cpp#L589
3. Since code in 1 & 2 are running on different threads (2 is executed on the main thread, 1 is executed on a newly created thread), and in some cases, the code in 1 runs faster, and it runs `controller->Destroy();` before the code in 2 runs `Reset();`
4. The controller destroys the controller, which, as a result, also destroys the debug adapter
5. Now the code 2 tries to run `Reset` on a now destroyed object, leading to the crash 

Note, the behavior of `DebuggerController::Destroy` deleting the DebuggerState (which then deletes the adapter) is newly introduces in https://github.com/Vector35/debugger/commit/67ded6c6f4c5134b38797460c538528291bfbaec#diff-b14c7886a752be824e0554606509430e92ed15e0b5edab2053abe9099d3f76f7R1180-R1184. I am not sure why I added this code
